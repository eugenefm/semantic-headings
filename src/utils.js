const getHeadingLevel = (auto, overide) => Math.min(overide || auto || 1, 6);

// eslint-disable-next-line import/prefer-default-export
export { getHeadingLevel };
