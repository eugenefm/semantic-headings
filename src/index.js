/* eslint-disable react/jsx-props-no-spreading */
import React, { createContext, forwardRef } from 'react';
import PropTypes from 'prop-types';
import { getHeadingLevel } from './utils';

const HeadingLevel = createContext(1);

export const SemanticSection = forwardRef(({ element: El, children, ...props }, ref) => (
  <HeadingLevel.Consumer>
    {(level) => (
      <HeadingLevel.Provider value={level + 1}>
        {El ? <El ref={ref} {...props}>{children}</El> : children}
      </HeadingLevel.Provider>
    )}
  </HeadingLevel.Consumer>
));

SemanticSection.propTypes = {
  element: PropTypes.elementType,
  children: PropTypes.node,
};

export const H = forwardRef(({ level, ...props }, ref) => (
  <HeadingLevel.Consumer>
    {(autoLevel) => {
      const HeadingElement = `h${getHeadingLevel(autoLevel, level)}`;
      return <HeadingElement ref={ref} {...props} />;
    }}
  </HeadingLevel.Consumer>
));

H.propTypes = {
  level: PropTypes.number,
};
