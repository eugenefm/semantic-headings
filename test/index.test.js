import React from 'react';
import { render } from '@testing-library/react';
import { H, SemanticSection } from '../src/index';
import '@testing-library/jest-dom/extend-expect';

describe('should render', () => {
  it('h1', () => {
    const { container } = render(<H>Hello World</H>);
    expect(container).toMatchSnapshot();
  });
  it('h2', () => {
    const { container } = render(<SemanticSection><H>Hello World</H></SemanticSection>);
    expect(container).toMatchSnapshot();
  });
  it('h2 wrapped in a <section>', () => {
    const { container } = render(<SemanticSection element="section"><H>Hello World</H></SemanticSection>);
    expect(container).toMatchSnapshot();
  });
  it('h1 through h3', () => {
    const { container } = render(
      <>
        <H>Hello World</H>
        <SemanticSection>
          <H>Hello World</H>
          <SemanticSection>
            <H>Hello World</H>
          </SemanticSection>
          <H>Hello World</H>
          <SemanticSection>
            <H>Hello World</H>
          </SemanticSection>
        </SemanticSection>
      </>,
    );
    expect(container).toMatchSnapshot();
  });
  it('h1 through h6 without going over', () => {
    const { container } = render(
      <>
        <H>Hello World</H>
        <SemanticSection>
          <H>Hello World</H>
          <SemanticSection>
            <H>Hello World</H>
            <SemanticSection>
              <H>Hello World</H>
              <SemanticSection>
                <H>Hello World</H>
                <SemanticSection>
                  <H>Hello World</H>
                  <SemanticSection>
                    <H>Hello World</H>
                    <SemanticSection>
                      <H>Hello World</H>
                    </SemanticSection>
                  </SemanticSection>
                </SemanticSection>
              </SemanticSection>
            </SemanticSection>
          </SemanticSection>
        </SemanticSection>
      </>,
    );
    expect(container).toMatchSnapshot();
  });
  it('h1 through h3 with override to h4', () => {
    const { container } = render(
      <>
        <H>Hello World</H>
        <SemanticSection>
          <H>Hello World</H>
          <SemanticSection>
            <H>Hello World</H>
          </SemanticSection>
          <H>Hello World</H>
          <SemanticSection>
            <H level={4}>Hello World</H>
          </SemanticSection>
        </SemanticSection>
      </>,
    );
    expect(container).toMatchSnapshot();
  });
});
