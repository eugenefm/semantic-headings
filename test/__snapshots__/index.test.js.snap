// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render h1 1`] = `
<div>
  <h1>
    Hello World
  </h1>
</div>
`;

exports[`should render h1 through h3 1`] = `
<div>
  <h1>
    Hello World
  </h1>
  <h2>
    Hello World
  </h2>
  <h3>
    Hello World
  </h3>
  <h2>
    Hello World
  </h2>
  <h3>
    Hello World
  </h3>
</div>
`;

exports[`should render h1 through h3 with override to h4 1`] = `
<div>
  <h1>
    Hello World
  </h1>
  <h2>
    Hello World
  </h2>
  <h3>
    Hello World
  </h3>
  <h2>
    Hello World
  </h2>
  <h4>
    Hello World
  </h4>
</div>
`;

exports[`should render h1 through h6 without going over 1`] = `
<div>
  <h1>
    Hello World
  </h1>
  <h2>
    Hello World
  </h2>
  <h3>
    Hello World
  </h3>
  <h4>
    Hello World
  </h4>
  <h5>
    Hello World
  </h5>
  <h6>
    Hello World
  </h6>
  <h6>
    Hello World
  </h6>
  <h6>
    Hello World
  </h6>
</div>
`;

exports[`should render h2 1`] = `
<div>
  <h2>
    Hello World
  </h2>
</div>
`;

exports[`should render h2 wrapped in a <section> 1`] = `
<div>
  <section>
    <h2>
      Hello World
    </h2>
  </section>
</div>
`;
